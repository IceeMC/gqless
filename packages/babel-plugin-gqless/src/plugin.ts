import jsxSyntax from '@babel/plugin-syntax-jsx'
import { types as t, PluginObj, BabelFileResult } from '@babel/core'
import { reactTransform } from './reactTransform'
import { importReferences } from './utils'
import { gqlessTransform } from './gqlessTransform'
import { Cache, FileAnalysis } from './analysis'

export type State = {
  cwd: string
  file: BabelFileResult
  filename: string

  analysis: FileAnalysis
}

export const plugin = (api: typeof import('@babel/core')): PluginObj<State> => {
  const cache = new Cache(api)

  return {
    name: 'gqless',
    inherits: jsxSyntax,
    pre() {
      const analysis = cache.getPath(this.filename)
      analysis.ast = this.file.ast!
      this.analysis = analysis
    },
    visitor: {
      ImportDeclaration(path, state) {
        switch (path.node.source.value) {
          case 'gqless': {
            importReferences(path, (importName, path) => {
              gqlessTransform(state.analysis, importName, path)
            })
            break
          }

          case '@gqless/react': {
            importReferences(path, (importName, path) => {
              reactTransform(importName, path)
            })
            break
          }
        }
      },
    },
  }
}
