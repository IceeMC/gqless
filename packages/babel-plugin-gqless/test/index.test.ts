// @ts-ignore
import pluginTester from 'babel-plugin-tester'
import plugin from '../src'

pluginTester({
  plugin,
  snapshot: true,
  tests: {
    'inserts name option for components': `
      import { graphql } from '@gqless/react'

      export const A = graphql(null)
      export const B = graphql(null, { existing: true })
      export const C = graphql(null, { name: 'Test' })
    `,

    'inserts name for useVariable': `
      import { useVariable } from '@gqless/react'

      const A = useVariable('')
      const B = useVariable('', 'Test')

      const C = useVariable('', false)
      const D = useVariable('', false, 'Test')
    `,

    'inserts name for useFragment': `
      import { useFragment } from '@gqless/react'

      const A = useFragment(null)
      const B = useFragment(null, 'User')
      const C = useFragment(null, 'User', 'Test')
    `,
  },
})
