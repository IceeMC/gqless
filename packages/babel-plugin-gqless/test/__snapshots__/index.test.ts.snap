// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gqless inserts name for useFragment: inserts name for useFragment 1`] = `
"
import { useFragment } from '@gqless/react'

const A = useFragment(null)
const B = useFragment(null, 'User')
const C = useFragment(null, 'User', 'Test')

      ↓ ↓ ↓ ↓ ↓ ↓

import { useFragment } from '@gqless/react';
const A = useFragment(null, undefined, \\"A\\");
const B = useFragment(null, 'User', \\"B\\");
const C = useFragment(null, 'User', 'Test');
"
`;

exports[`gqless inserts name for useVariable: inserts name for useVariable 1`] = `
"
import { useVariable } from '@gqless/react'

const A = useVariable('')
const B = useVariable('', 'Test')

const C = useVariable('', false)
const D = useVariable('', false, 'Test')

      ↓ ↓ ↓ ↓ ↓ ↓

import { useVariable } from '@gqless/react';
const A = useVariable('', \\"A\\");
const B = useVariable('', 'Test');
const C = useVariable('', false, \\"C\\");
const D = useVariable('', false, 'Test');
"
`;

exports[`gqless inserts name option for components: inserts name option for components 1`] = `
"
import { graphql } from '@gqless/react'

export const A = graphql(null)
export const B = graphql(null, { existing: true })
export const C = graphql(null, { name: 'Test' })

      ↓ ↓ ↓ ↓ ↓ ↓

import { graphql } from '@gqless/react';
export const A = graphql(null, {
  name: \\"A\\"
});
export const B = graphql(null, {
  name: \\"B\\",
  existing: true
});
export const C = graphql(null, {
  name: 'Test'
});
"
`;
